internal/node/telemetry.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	err := t.init()
+	if err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *Telemetry) init() error {
+	var e, err error
+
+	e = t.initMeterProvider()
+	if err != nil {
+		err = errors.Join(err, e)
+	}
+
+	if t.meter == nil {
+		return errors.New("failed to initialize telemetry instance: nil meter")
+	}
+
+	t.vmCounter, e = t.meter.
+		Int64UpDownCounter("nex-vm-count",
+			metric.WithDescription("Number of VMs started"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.workloadCounter, e = t.meter.
+		Int64UpDownCounter("nex-workload-count",
+			metric.WithDescription("Number of workloads deployed"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.deployedByteCounter, e = t.meter.
+		Int64UpDownCounter("nex-deployed-bytes-count",
+			metric.WithDescription("Total number of bytes deployed"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.allocatedMemoryCounter, e = t.meter.
+		Int64UpDownCounter("nex-total-memory-allocation-mib-count",
+			metric.WithDescription("Total allocated memory based on firecracker config"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.allocatedVCPUCounter, e = t.meter.
+		Int64UpDownCounter("nex-total-vcpu-allocation-count",
+			metric.WithDescription("Total allocated VCPU based on firecracker config"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+
+	t.functionTriggers, e = t.meter.
+		Int64Counter("nex-function-trigger",
+			metric.WithDescription("Total number of times a function was triggered"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.functionFailedTriggers, e = t.meter.
+		Int64Counter("nex-function-failed-trigger",
+			metric.WithDescription("Total number of times a function failed to triggered"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	t.functionRunTimeNano, e = t.meter.
+		Int64Counter("nex-function-runtime-nanosec",
+			metric.WithDescription("Total run time in nanoseconds for function"),
+		)
+	if e != nil {
+		err = errors.Join(err, e)
+	}
+	return err
+}
+
+func (t *Telemetry) initMeterProvider() error {
+	resource, err := resource.Merge(resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceName(t.serviceName),
+			semconv.ServiceVersion(VERSION),
+		))
+
+	if err != nil {
+		t.log.Warn("failed to create OTel resource", slog.Any("err", err))
+		return err
+	}
+
+	metricReader, err := t.serveMetrics()
+	if err != nil {
+		t.log.Warn("failed to create OTel metrics exporter", slog.Any("err", err))
+		return err
+	}
+
+	meterProvider := metricsdk.NewMeterProvider(
+		metricsdk.WithResource(resource),
+		metricsdk.WithReader(
+			metricReader,
+		),
+	)
+
+	defer func() {
+		if err := meterProvider.Shutdown(context.Background()); err != nil {
+			t.log.Error("failed to shutdown OTel meter provider", slog.Any("err", err))
+		}
+	}()
+
+	otel.SetMeterProvider(meterProvider)
+	return nil
+}
+
+func (t *Telemetry) serveMetrics() (metricsdk.Reader, error) {
```

Comment:
we need to talk about moving this out of Telemetry.

Reply:
I added a comment for us to do this. I vote to merge this one and tackle changes to `serveMetrics()` in another PR. I presume the idea is this feels a lot better in `Node` and stopped on shutdown.